### PR TITLE
Swift 3

### PR DESCRIPTION
**Pull request by mikeger**
- God bless auto converter

### CODE REVIEW

File: ziphy.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -487,6 +491,8 @@
 				PRODUCT_BUNDLE_IDENTIFIER = "com.wearezeta.$(PRODUCT_NAME:rfc1034identifier)";
 				PRODUCT_NAME = "$(TARGET_NAME)";
 				SKIP_INSTALL = YES;
+				SWIFT_TREAT_WARNINGS_AS_ERRORS = YES;
```
Comment (blender) [inline ziphy.xcodeproj/project.pbxproj]: Sure about this one?

Comment (mikeger) [inline ziphy.xcodeproj/project.pbxproj]: Yes.
znakcom-37805-600x416

File: ziphy/Ziph.swift [hunk 2]
```diff
@@ -19,12 +19,12 @@
 
 import Foundation
 
-@objc public class Ziph : NSObject {
+@objc open class Ziph : NSObject {
```
Comment (daehn) [inline ziphy/Ziph.swift]: Do e ever subclass a `Ziph`? If not this could be `public`.

File: ziphy/ZiphyClient.swift [hunk 3]
```diff
@@ -23,30 +23,30 @@ import Foundation
...
+private let searchEndpoint = gifsEndpoint + "/search"
+private let randomEndpoint = gifsEndpoint + "/random"
 private let requestScheme = "https"
 
 
 
-public typealias ZiphsCallBack = (success:Bool, ziphs:[Ziph], error:NSError?) -> ()
-public typealias ZiphByIdCallBack = (success:Bool, ziphId:String, error:NSError?)->()
-public typealias ZiphyImageCallBack = (success:Bool, image:ZiphyImageRep?, ziph:Ziph, data:NSData?, error:NSError?) -> ()
+public typealias ZiphsCallBack = (_ success:Bool, _ ziphs:[Ziph], _ error:Error?) -> ()
```
Comment (blender) [inline ziphy/ZiphyClient.swift]: Sure you want to erase all named parameters ?

File: ziphy/ZiphyClient.swift [hunk 4]
```diff
@@ -85,63 +85,63 @@ public typealias ZiphyImageCallBack = (success:Bool, image:ZiphyImageRep?, ziph:
                     return eitherImageArray.rightMap { ziphs in
                         
                         performOnQueue(callBackQueue) {
-                            onCompletion(success: true, ziphs: ziphs, error: nil)
+                            onCompletion(true, ziphs, nil)
                         }
```
Comment (blender) [inline ziphy/ZiphyClient.swift]: Doesn't this look a little voodoo without labels?

File: ziphy/ZiphyImageFetcher.swift [hunk 5]
```diff
@@ -18,31 +18,51 @@
...
 private let giphyApiHost = "api.giphy.com"
 
 
-public typealias SingleZiphCallBack = (imageData:NSData?, ziph:Ziph?,  error:NSError?) -> ()
+public typealias SingleZiphCallBack = (_ imageData:Data?, _ ziph:Ziph?,  _ error:Error?) -> ()
 
 
-@available(*, deprecated=2.0)
-@objc public class ZiphyImageFetcher: NSObject {
+@available(*, deprecated: 2.0)
```
Comment (blender) [inline ziphy/ZiphyImageFetcher.swift]: This guy should go now.

File: ziphy/Ziph.swift (hunk 2 above)
Comment (blender) [inline ziphy/Ziph.swift]: There is really no reason why one would want to subclass ziph

File: ziphy/ZiphyImageRep.swift [hunk 6]
```diff
@@ -20,16 +20,16 @@
 import Foundation
 
 
-@objc public class ZiphyImageRep:ZiphyStillImageRep {
+@objc open class ZiphyImageRep:ZiphyStillImageRep {
```
Comment (blender) [inline ziphy/ZiphyImageRep.swift]: Also this maybe public and not open

File: ziphy/ZiphyRequestGenerator.swift [hunk 7]
```diff
@@ -29,68 +29,68 @@ struct ZiphyRequestGenerator {
     let randomEndpoint:String
     let gifsEndpoint:String
     
-    private func requestWithParameters(endPoint:String, query:String? = nil) -> Either<NSError, NSURLRequest> {
+    fileprivate func requestWithParameters(_ endPoint:String, query:String? = nil) -> Either<Error, URLRequest> {
         
-        let components = NSURLComponents()
+        var components = URLComponents()
```
Comment (blender) [inline ziphy/ZiphyRequestGenerator.swift]: Var?

File: ziphy/ZiphyUtils.swift [hunk 8]
```diff
@@ -83,45 +83,45 @@ enum Either<L, R> {
...
+    func identity<A>(_ a: A) -> A {
         return a
     }
 
 }
 
-typealias NSURLRequestCallBack = (data:NSData?, response:NSURLResponse?, error:NSError?)->NSError?
+typealias NSURLRequestCallBack = (_ data:Data?, _ response:URLResponse?, _ error:Error?)->Error?
 
 class NSURLRequestPromise {
```
Comment (blender) [inline ziphy/ZiphyUtils.swift]: should drop NS prefix

File: ziphy/ZiphyUtils.swift [hunk 9]
```diff
@@ -83,45 +83,45 @@ enum Either<L, R> {
...
     
-    func identity<A>(a: A) -> A {
+    func identity<A>(_ a: A) -> A {
         return a
     }
 
 }
 
-typealias NSURLRequestCallBack = (data:NSData?, response:NSURLResponse?, error:NSError?)->NSError?
+typealias NSURLRequestCallBack = (_ data:Data?, _ response:URLResponse?, _ error:Error?)->Error?
```
Comment (blender) [inline ziphy/ZiphyUtils.swift]: Drop NS prefix

File: ziphyTests/ZiphyPaginationControllerTests.swift [hunk 10]
```diff
@@ -101,36 +101,31 @@ class ZiphyPaginationControllerTests: ZiphyTestCase {
...
         
-        //There are only 13 results for this search term so this should not trigger more than 2 requests of 10 each
         self.paginationController.fetchBlock = self.fetchBlockForSearch(self.paginationController, searchTerm: "awg", resultsLimit: 10)
         self.paginationController.completionBlock = { [weak self](success, error) in
             
-            if (success && self?.paginationController.totalPagesFetched < 3) {
+            if (success) {
                 self?.paginationController.fetchNewPage()
             }
-            else if (success && self?.paginationController.totalPagesFetched == 3) {
```
Comment (blender) [inline ziphyTests/ZiphyPaginationControllerTests.swift]: What happened here? This is a logic change

File: ziphy/Ziph.swift (hunk 2 above)
Comment (daehn) [inline ziphy/Ziph.swift]: Unfortunately it seems like the migrator replaces all `public` declarations with `open`, I would suggest using `public` by default as there a a couple of places where this happened.

File: ziphyTests/ZiphyPaginationControllerTests.swift (hunk 10 above)
Comment (mikeger) [inline ziphyTests/ZiphyPaginationControllerTests.swift]: Test implied "There are only 13 results for this search" but it is not true any more and would change in the future, so I removed the part of the test that was based on amount of replies.

**Comment by daehn** [comment]
On another note, it seems we didn't setup Travis for this project yet, as they already have a Xcode 8 GM image it would be nice to add it.

File: ziphy/Ziph.swift (hunk 2 above)
Comment (mikeger) [inline ziphy/Ziph.swift]: That's true, autoconvert does open per default. Unfortunately the converter has only one button (Convert!) and no settings.

File: ziphy/ZiphyRequestGenerator.swift (hunk 7 above)
Comment (mikeger) [inline ziphy/ZiphyRequestGenerator.swift]: Swift 3 also marks as var the objects that you modify (see lines 36-37)

File: ziphyTests/ZiphyPaginationControllerTests.swift (hunk 10 above)
Comment (blender) [inline ziphyTests/ZiphyPaginationControllerTests.swift]: Ok. However you might want to test (somewhere else probably) that there is a case where .NoMorePages is actually fired.

**Comment by mikeger** [comment]
@daehn there is a travis integration branch, it was not merged unfortunately
https://github.com/wireapp/wire-ios-ziphy/compare/feature/travis-ci?expand=1

I would re-do Travis integration as a next PR

File: ziphy/ZiphyRequestGenerator.swift (hunk 7 above)
Comment (blender) [inline ziphy/ZiphyRequestGenerator.swift]: ok.

File: ziphy/ZiphyStillImageRep.swift [hunk 11]
```diff
@@ -20,15 +20,15 @@
 import Foundation
 
 
-@objc public class ZiphyStillImageRep: NSObject {
+@objc open class ZiphyStillImageRep: NSObject {
```
Comment (blender) [inline ziphy/ZiphyStillImageRep.swift]: here again probably not `open` but `public`

**Comment by blender** [comment]
Looks good to me apart from removing the labels from the callbacks which in my opinion reads are voodoo.

File: ziphy/ZiphyClient.swift (hunk 3 above)
Comment (mikeger) [inline ziphy/ZiphyClient.swift]: In Swift 3 the arguments labels for closure are not allowed: https://github.com/apple/swift-evolution/blob/master/proposals/0111-remove-arg-label-type-significance.md
Another reference: https://lists.swift.org/pipermail/swift-evolution-announce/2016-July/000233.html

One of the most annoying changes.

File: ziphy/ZiphyClient.swift (hunk 4 above)
Comment (mikeger) [inline ziphy/ZiphyClient.swift]: It is.

**Comment by mikeger** [comment]
@daehn @blender updated

File: ziphy/Ziph.swift [hunk 12]
```diff
@@ -19,7 +19,7 @@
 
 import Foundation
 
-@objc open class Ziph : NSObject {
+@objc public class Ziph : NSObject {
     
     open let ziphId:String!
```
Comment (blender) [inline ziphy/Ziph.swift]: I guess this is ok? Class public but properties open?

File: ziphy/ZiphyClient.swift (hunk 3 above)
Comment (blender) [inline ziphy/ZiphyClient.swift]: ok 

seems very clear now.
```
doSomething(nil, .Nothing, "", false, nil, callback1, true ,callback2)
```

**Comment by blender** [comment]
Fine with me

**Pull request merged**